### ISSUE TITLE
Segmentation Faults upon logging into wp-admin 

### ISSUE THREAD

**Issue by jayhill90** [woocommerce/woocommerce#23352]
**Describe the bug**
When using WP Engine and Woocommerce 3.6.0, logging into wp-admin results in a 502 due to a segmentation fault. 

**To Reproduce**
1. Login to https://testinproduction.systems via my-account or wp-login with <redacted> as credentials 
2. 502 will be presented.

In the Apache error logs from WP Engine, we see ` child pid 14937 exit signal Segmentation fault (11)` along with the GET request for `/wp-admin/`

```
MYIP testinprod.wpengine.com - [17/Apr/2019:18:29:49 +0000] "GET /wp-admin/ HTTP/1.1" 502 1175 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_4) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/73.0.3683.103 Safari/537.36"
```
This issue does not occur in another environment that I have setup on a Digital Ocean droplet also running PHP 7.2

The segmentation fault also occurs if logging out. 

**Expected behavior**

**Isolating the problem (mark completed items with an [x]):**
- [X] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [X] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [X] I can reproduce this bug consistently using the steps above.

**WordPress Environment**
PHP 7.2. This doesn't occur in PHP 7.3 though. 
WordPress 5.1.1
```

WordPress address (URL): |   | https://testinproduction.systems
-- | -- | --
Site address (URL): |   | https://testinproduction.systems
WooCommerce version: |   | 3.6.0
Log directory writable: |   | /nas/content/live/testinprod/wp-content/uploads/wc-logs/
WordPress version: |   | 5.1.1
WordPress multisite: |   | –
WordPress memory limit: |   | 512 MB
WordPress debug mode: |   | –
WordPress cron: |   | –
Language: |   | en_US
External object cache: |   |  
```
</details>

Was there anything in 3.6.0 related to logging in that was changed? I didn't see much on the changelog that would potentially affect this.

**Comment by itGuyFacethesound** [comment]
It happen on my install too, also wpengine, also it works in local, i've opened a ticket also on wpengine side.

**Comment by rrennick** [comment]
@jayhill90 Please don't post login credentials in an issue (I removed the credentials). These are public and indexed by search engines.

**Comment by claudiosanches** [comment]
@jayhill90 probably good to remove that user from your installation, since it's still possible to check those, even after @rrennick edited.

**Comment by jayhill90** [comment]
It was a test account with no role or permissions but thanks. Just wanted y’all to be able to replicate if needed.

**Comment by GlennMartin1** [comment]
Front end works, but cannot log into back end.

On all sites with:
WP 5.1.1
WC 3.6.0
WP Engine
PHP 7.2

**Comment by jayhill90** [comment]
Was there anything login related that changed in 3.6.0?

**Comment by aussio** [comment]
Hey folks, I have just a tiny amount more information from poking around on a WP Engine site with 3.6.8.

I'm seeing the code exiting at [this point in the code](https://github.com/WordPress/WordPress/blob/master/wp-login.php#L1001-L1002). I added break points before the `wp_redirect`, before the `exit()`, and after the `exit()`. I can confirm the `exit()` is being called and then the code is not continuing (the breakpoint after `exit()` isn't reached as expected).

I'm seeing a Segmentation Fault in our Apache error logs. It represents it as a 502 in the WP Engine stack due to Nginx proxying to Apache. 

* I cannot reproduce on 3.5.8. I flip flopped back and forth a good few times and could reproduce every time.
* The $_SERVER variable looks entirely the same
* The Location and Status headers set by wp_redirect there are exactly the same on 3.5.8 and 3.6.0.

Hopefully this is helpful. I'm unsure of what else to look into at the moment. :)

**Comment by mikejolley** [comment]
We've traced it down to this line: https://github.com/woocommerce/woocommerce/blob/b7971bd0dcad2ef9bd9fd39f711eff6d941e1497/includes/class-wc-cart-session.php#L56-L57

Why this fails under WPE 7.2 we're not yet sure but changing this block from:

```
if ( function_exists( 'header_register_callback' ) ) {
			header_register_callback( array( $this, 'maybe_set_cart_cookies' ) ); // phpcs:ignore PHPCompatibility.FunctionUse.NewFunctions.header_register_callbackFound
		} else {
			add_action( 'woocommerce_add_to_cart', array( $this, 'maybe_set_cart_cookies' ) );
			add_action( 'wp', array( $this, 'maybe_set_cart_cookies' ), 99 );
			add_action( 'shutdown', array( $this, 'maybe_set_cart_cookies' ), 0 );
		}
```

to:

```
add_action( 'woocommerce_add_to_cart', array( $this, 'maybe_set_cart_cookies' ) );
add_action( 'wp', array( $this, 'maybe_set_cart_cookies' ), 99 );
add_action( 'shutdown', array( $this, 'maybe_set_cart_cookies' ), 0 );
```

will workaround the issue until addressed by WPE.

**Comment by mikejolley** [comment]
@aussio I did not realise you were from WPE :) See my comment above ^ I'm not sure if it's specific to the version of PHP 7.2 being ran (maybe it has been patched?), but a quick fix may be to disable that function for users running that PHP version if an upgrade is not possible?

**Comment by aussio** [comment]
Thanks! I'm trying to sort out why we're tripping up on that if-statement right now. Thanks so much for helping out here. :)

**Comment by peterfabian** [comment]
@aussio it's probably the `header_register_callback` function itself. There are a few seg faults reported in different versions of PHP it seems, by looking at bug trackers.

**Comment by timmyc** [comment]
Hey all watching this issue, wanted to close the loop and let you all know the fix has been shipped in 3.6.1: https://woocommerce.wordpress.com/2019/04/18/woocommerce-3-6-1-release/

**Comment by jayhill90** [comment]
Thanks a ton for the quick turn around!
